### PR TITLE
 JBPM-8108:Improve zoom action on process diagram 

### PR DESCRIPTION
**Pull request by nmirasch**

### CODE REVIEW

**Comment by nmirasch** [comment]
depends on https://github.com/kiegroup/appformer/pull/608 and includes the commit in https://github.com/kiegroup/jbpm-wb/pull/1259.
Just to start a discussion about what's the best approach for the zoom behaviour

@cristianonicolai not sure if here is the best way to fix the height of the diagram region to the max available in the window. Any idea for improve that?
@cristianonicolai @jomarko could you take a look?  Thanks!

**Comment by cristianonicolai** [comment]
Jenkins execute full downstream build

**Comment by cristianonicolai** [comment]
Jenkins execute full downstream build

**Comment by nmirasch** [comment]
Following @jomarko suggestions I have added some test and disable the  + and - buttons when the zoom limits have been reached. 
I have added the same header when the diagram is shown at process definition details  as is shown at process instance details
@cristianonicolai @jomarko could you take a look? Thanks!

**Comment by cristianonicolai** [comment]
Jenkins execute full downstream build

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/definition/details/ProcessDefinitionDetailsPresenter.java [hunk 1]
```diff
@@ -160,7 +163,8 @@ public IsWidget getDetailsView() {
     }
 
     public IsWidget getProcessDiagramView() {
-        return processDiagramPresenter.getView();
+        processDefinitionDiagramView.setDiagram(processDiagramPresenter.getView());
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/definition/details/ProcessDefinitionDetailsPresenter.java]: @nmirasch do you think this view could be moved to the presenter, this looks a bit strange to me. Maybe refactor a bit the common bits between definition and instance diagrams if needed.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java [hunk 2]
```diff
@@ -53,6 +53,12 @@
 
     private Callback<String> nodeSelectionCallback;
 
+    private D3 d3Builded = D3.Builder.get();
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java]: Perhaps use a `@PostConstruct` method to initialize.

Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java]: maybe just call it `d3` ?

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java [hunk 3]
```diff
@@ -63,16 +69,33 @@ public void displayImage(final String svgContent) {
 
         processDiagramDiv.innerHTML = svgContent;
 
-        final D3 d3 = D3.Builder.get();
+        final D3 d3 = d3Builded;
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java]: no need, just reuse same variable.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java [hunk 4]
```diff
@@ -113,6 +113,22 @@ public HTMLElement getElement() {
...
+    public void disableMinusButton(boolean disabled) {
+        disableButton(scaleMinus, disabled);
+    }
+
+    public void disableButton(HTMLAnchorElement anchorElement, boolean disabled) {
+        String buttonCssClass = "btn btn-default";
+        if (disabled) {
+            buttonCssClass += " disabled";
+        }
+        anchorElement.setAttribute("class", buttonCssClass);
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java]: You can manipulate the class list directly using: `anchorElement.classList.add();` and  `anchorElement.classList.remove();`

Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java]: no need to set `disabled=true`?

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/definition/details/ProcessDefinitionDetailsPresenter.java (hunk 1 above)
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/definition/details/ProcessDefinitionDetailsPresenter.java]: oks

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java (hunk 2 above)
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewImpl.java]: done

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java (hunk 4 above)
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java]: oks

Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ZoomControlView.java]: I don't think it's needed with the disable class it doesn't react to pointer events 
a.btn.disabled, fieldset[disabled] a.btn {
    pointer-events: none;
}

**Comment by nmirasch** [comment]
@cristianonicolai Thanks!  changed a bit the code to add your suggestions. Could you give it another try?

**Comment by nmirasch** [comment]
Jenkins execute full downstream build

**Comment by jomarko** [comment]
@nmirasch thanks for added tests, just ping me when downstream build finish successfully and I will do last manual check.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramPresenter.java [hunk 5]
```diff
@@ -63,6 +63,11 @@ public void displayImage(final String svgContent, final String containerId) {
         view.hideBusyIndicator();
     }
 
+    public void setViewExpandedMode() {
+        view.removeDiagramContainerClass("col-md-10");
```
Comment (cristianonicolai) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramPresenter.java]: hm, presenter dealing with css classes?

Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramPresenter.java]: ups, fixing it

**Comment by nmirasch** [comment]
@cristianonicolai done

**Comment by nmirasch** [comment]
@cristianonicolai I have add a change on the diagram container heigh calculation to avoid that vertical scroll at business central.

**Comment by cristianonicolai** [comment]
Jenkins execute full downstream build

**Comment by cristianonicolai** [comment]
@jomarko should be ready for review now.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewTest.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    private D3 d3Mock;
+
+    @Mock
+    private D3 svgSelect;
+
+    @Mock
+    private D3.Zoom zoomMock;
+
+    @Mock
+    D3.Transform transformMock;
```
[the remark this answers is not in the document]
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewTest.java]: done

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewTest.java [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(GwtMockitoTestRunner.class)
+public class ProcessDiagramWidgetViewTest {
+
+    @InjectMocks
+    ProcessDiagramWidgetViewImpl view;
```
[the remark this answers is not in the document]
Comment (nmirasch) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/test/java/org/jbpm/workbench/pr/client/editors/diagram/ProcessDiagramWidgetViewTest.java]: done

**Pull request merged**